libcache/cache.c

```diff
@@ -0,0 +1,650 @@
+{
+	int ret = 0;
+
+	cacheline_t **lhsLine = (cacheline_t **)lhs;
+	cacheline_t **rhsLine = (cacheline_t **)rhs;
+
+	if (*lhsLine == NULL && *rhsLine == NULL) {
+		ret = 0;
+	}
+	else if (*lhsLine == NULL && *rhsLine != NULL) {
+		ret = -1;
+	}
+	else if (*lhsLine != NULL && *rhsLine == NULL) {
+		ret = 1;
+	}
+	else {
+		if ((*lhsLine)->tag < (*rhsLine)->tag) {
+			ret = -1;
+		}
+		else if ((*lhsLine)->tag == (*rhsLine)->tag) {
+			ret = 0;
+		}
+		else {
+			ret = 1;
+		}
+	}
+
+	return ret;
+}
+
+
+static ssize_t cache_flushLine(cache_writeCb_t writeCb, cacheline_t *linePtr, const uint64_t addr, size_t count)
+{
+	ssize_t written = 0;
+
+	if (linePtr != NULL && IS_VALID(linePtr->flags) && IS_DIRTY(linePtr->flags)) {
+		written = (*writeCb)(addr, linePtr->data, sizeof(*linePtr->data), count);
+		CLEAR_DIRTY(linePtr->flags);
+		if (written != count) {
+			written = -1;
+		}
+	}
+
+	return written;
+}
+
+
+static ssize_t cache_executePolicy(cache_writeCb_t writeCb, cacheline_t *linePtr, const uint64_t addr, size_t count, int policy)
+{
+	ssize_t written = 0;
+
+	switch (policy) {
+		case 0:
+			/* write-back */
+			SET_DIRTY(linePtr->flags);
+			written = count;
+			break;
+		case 1:
+			/* write-through */
+			written = (*writeCb)(addr, linePtr->data, sizeof(*linePtr->data), count);
+			break;
+		default:
+			break;
+	}
+
+	return written;
+}
+
+
+static ssize_t cache_writeToSet(cache_writeCb_t writeCb, cacheset_t *set, cacheline_t *line, const uint64_t offset, size_t count, int policy)
+{
+	int i;
+	ssize_t written = 0;
+	cacheline_t *temp = NULL;
+
+	if (set->count < LIBCACHE_NUM_WAYS) {
+		set->count++;
+
+		for (i = 0; i < set->count; ++i) {
+			if (!IS_VALID(set->lines[i].flags)) {
+				set->lines[i] = *line;
+				break;
+			}
+		}
+
+		temp = &(set->lines[i]);
+
+		for (i = 0; i < set->count; ++i) {
+			if (set->tags[i] == NULL) {
+				set->tags[i] = temp;
+				break;
+			}
+		}
+
+		LIST_ADD(&set->timestamps, temp);
+	}
+	else {
+		temp = set->timestamps;
+		LIST_REMOVE(&set->timestamps, temp);
+
+		/* write-back */
+		if (IS_DIRTY(temp->flags)) {
+			written = cache_flushLine(writeCb, temp, offset, count);
+			if (written == -1) {
+				return -1;
+			}
+			free(temp->data);
+		}
+
+		for (i = 0; i < set->count; ++i) {
+			if (IS_VALID(set->lines[i].flags) && (set->lines[i].tag == temp->tag)) {
+				set->lines[i] = *line;
+				break;
+			}
+		}
+
+		LIST_ADD(&set->timestamps, temp);
+	}
+
+	written = cache_executePolicy(writeCb, temp, offset, count, policy);
+
+	qsort(set->tags, LIBCACHE_NUM_WAYS, sizeof(cacheline_t *), cache_compareTags);
+
+	return written;
+}
+
+
+static cacheline_t *cache_findLine(cacheset_t *set, uint64_t tag)
+{
+	cacheline_t key, *keyPtr = &key, **linePtr = NULL;
+
+	key.tag = tag;
+
+	linePtr = bsearch(&keyPtr, set->tags, LIBCACHE_NUM_WAYS, sizeof(cacheline_t *), cache_compareTags);
+
+	return (linePtr != NULL) ? *linePtr : NULL;
+}
+
+
+static cacheline_t *cache_updateTimestamps(cacheset_t *set, uint64_t tag)
```

Comment:
Name of this function is misleading - it's main purpose is really to find a line and used so.
imho we should merge cache_findLine and cache_updateTimestamps functions together - so find line will update timestamp, as we access line every time it's found

Reply:
The name is unfortunate indeed. I've merged and rearranged the functions and it has solved the next problem (with 2 searches). Thank you